### PR TITLE
Add mingw-w64 CI using msys2

### PR DESCRIPTION
**Pull request by jeroen**
Adds CI for mingw-w64 using the awesome [msys2 action](https://github.com/msys2/setup-msys2). I adapted the build configuration and patches from upstream msys2 [mingw-w64-imagemagick](https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-imagemagick).

I think this may be useful to debug some issues and perhaps fix some compiler warnings.

Update: I commented out `make check` because it takes very long.

### CODE REVIEW

File: MagickCore/nt-base.c [hunk 1]
```diff
@@ -1800,7 +1800,14 @@ MagickPrivate DIR *NTOpenDirectory(const char *path)
     MagickPathExtent);
   if (length == 0)
     return((DIR *) NULL);
+#ifdef __MINGW32__
+  wchar_t WCDirectorySeparator[strlen(DirectorySeparator)+1];
```
Comment (dlemstra) [inline MagickCore/nt-base.c:1804]: This indeed incorrect but the solution is actually to remove appending the backslash. This is written in the MSDN documentation:

> This parameter should not be NULL, an invalid string (for example, an empty string or a string that is missing the terminating null character), or end in a trailing backslash (\).

I just pushed a patch to resolve this.

File: MagickCore/nt-base.h [hunk 2]
```diff
@@ -29,7 +29,11 @@ extern "C" {
 
 #define WIN32_LEAN_AND_MEAN
 #define VC_EXTRALEAN
+#ifndef __MINGW32__
 #define _CRT_SECURE_NO_DEPRECATE  1
```
Comment (dlemstra) [inline MagickCore/nt-base.h:33]: Why is this added?

File: MagickCore/nt-base.h [hunk 3]
```diff
@@ -99,7 +103,9 @@ extern "C" {
 #if !defined(closedir)
 #  define closedir(directory)  NTCloseDirectory(directory)
 #endif
-#define MAGICKCORE_HAVE_ERF
+#if !defined(MAGICKCORE_HAVE_ERF)
```
Comment (dlemstra) [inline MagickCore/nt-base.h:106]: Why do we need this?

**Review by dlemstra — COMMENTED** [review]
Thanks for your contribution. It is really a good idea to also test the mingw compilation. I left some comments/questions.

File: .github/workflows/mingw-w64-imagemagick/PKGBUILD [hunk 4]
```diff
@@ -0,0 +1,141 @@
+# Adapted from upstream:
```
[the remark this answers is not in the document]
Comment (jeroen) [reply .github/workflows/mingw-w64-imagemagick/PKGBUILD:1]: I tried to keep the upstream msys2 naming convention, you could imagine having multiple build flavours each with their own PKGBUILD. But I suppose we could move it into the root if you prefer.

File: .github/workflows/mingw-w64-imagemagick/PKGBUILD [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+options=('staticlibs' 'strip' '!debug' 'libtool')
+source_dir="$GITHUB_WORKSPACE"
+validpgpkeys=('D8272EF51DA223E4D05B466989AB63D48277377A')
+
+prepare() {
+  cd ${source_dir}
+  # autoreconf -fi
+}
+
+build() {
```
[the remark this answers is not in the document]
Comment (jeroen) [reply .github/workflows/mingw-w64-imagemagick/PKGBUILD:62]: I think it is best to stay close to the upstream PKGBUILD: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-imagemagick/PKGBUILD

That way we test the entire process and dependencies automatically get resolved, and so on.

File: MagickCore/distribute-cache.c [hunk 6]
```diff
@@ -51,6 +51,10 @@
 /*
   Include declarations.
 */
+#if defined(MAGICKCORE_WINDOWS_SUPPORT)
+#include <winsock2.h>
+#include <ws2tcpip.h>
```
[the remark this answers is not in the document]
Comment (jeroen) [reply MagickCore/distribute-cache.c:56]: I have copied this from msys2, but perhaps it is no longer needed: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-imagemagick/002-7.0.4.1-build-fixes.patch

Generally, mingw-w64 is more convervative with which headers get included by default, so it could be that this is actually needed. You could try removing it and see if it still builds...

File: MagickCore/nt-base.h (hunk 2 above)
Comment (jeroen) [reply MagickCore/nt-base.h:33]: Otherwise you get million "_CRT_SECURE_NO_DEPRECATE redefined" warnings in the build. I suppose mingw-w64 defines this by default

File: MagickCore/nt-base.h (hunk 3 above)
Comment (jeroen) [reply MagickCore/nt-base.h:106]: Same as above, it is defined by default in mingw-w64 (perhaps via configure)?

**Comment by jeroen** [comment]
Thanks for your comments. Feel free to push changes into this branch to test what happens in CI.

File: MagickCore/nt-base.h (hunk 2 above)
Comment (dlemstra) [reply MagickCore/nt-base.h:33]: I have applied these changes with a separate commit.

File: MagickCore/nt-base.h (hunk 3 above)
Comment (dlemstra) [reply MagickCore/nt-base.h:106]: I have applied these changes with a separate commit.

**Comment by jeroen** [comment]
I have rebased the PR and incorporated your suggestions

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -113,3 +113,28 @@ jobs:
...
+      matrix:
+        include: [
+          { msystem: ucrt64, toolchain: ucrt-x86_64 },
+          { msystem: mingw64, toolchain: x86_64 },
+          { msystem: mingw32, toolchain: i686 }
+        ]
+      fail-fast: false
+    steps:
+      - name: Prepare git
+        run: git config --global core.autocrlf false
```
[the remark this answers is not in the document]
Comment (jeroen) [reply .github/workflows/main.yml:130]: Git automatically converts line endings to Windows format, but the PKGBUILD file should remain in unix format.

[the remark this answers is not in the document]
Comment (jeroen) [reply .github/workflows/main.yml:130]: An alternative solution would be to set this in .gitattributes but then it might also affect other systems.

**Comment by yg8ijvjvjv** [comment]
Why would ImageMagick add a copyright infringement of MinGW-w64?

**Comment by dlemstra** [comment]
Please revert the commit to store the artifacts @jeroen, we don't want to create artifacts every commit.

And I don't understand what you are trying to tell us @yg8ijvjvjv

**Comment by jeroen** [comment]
@dlemstra ok I have removed that and rebased again
